### PR TITLE
Continue line comments

### PR DESCRIPTION
**Pull request by TornaxO7**
Closes https://github.com/helix-editor/helix/issues/1730

Continues https://github.com/helix-editor/helix/pull/8664

This is my attempt to apply the changes of https://github.com/helix-editor/helix/pull/4718 to the PR.

### CODE REVIEW

**Comment by TornaxO7** [comment]
One thing which bothers me is how we want to "continue" the comment from block comments like for C of this style:
```
/*
 * | <-- let's say here's the cursor, how should we "recognize" this `*`? Adding another entry in `language.toml` per language which should indicate which "multiblock-continuity" there are?
 */
```

**Review by pascalkuthe — COMMENTED** [review]
Takes very little code to get this working :+1: 

I wonder if instead of using a plaintext detection we should be using tree sitter. We already have comment text objects. This would prevent continue comments in string literals (and maybe some niche cases in some langauges). The again c-c does not use tree sitter either so maybe it's not necessary. We can probably delay it to future work and and then update all comment functionality to use TS together.

**Comment by pascalkuthe** [comment]
regarding block comments the continue token (like *) should be a seperate field in the `BlockCommentToken` named `continue` (or similar). Then you would need to populate this field for all languages. By applying my comments you will likely find out "for free" if you are inside the block comment by reusing existing functions. I don't see it as a must for this PR but definitly would be nice to have

**Comment by TornaxO7** [comment]
> I don't see it as a must for this PR but definitly would be nice to have

I think, I'd like to create a followup PR for block comments.

**Comment by TornaxO7** [comment]
Ok, to sum up the status:

1. I've implemented the continue-comment feature *only* for line-comments now because I think it's fine to create a follow-up PR for block comments and I don't want to make this PR too big.
2. Auto-comment-continueing is enabled per default
3. I'm using Hybrid-Indent detection at the moment for new line with comments. I hope that this is fine?

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:3395]: The rename here isn't necessary, can be reset

File: helix-view/src/editor.rs
Comment (the-mikedavis) [inline helix-view/src/editor.rs:350]:
```suggestion
```

This doesn't need a default directive for serde since `Default for Config` is implemented manually.

Also the doc here is outdated that the default is false. This should also be added to the book

I wonder if this needs to be configurable at all? I'm not sure it's configurable in Kakoune for example

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:3435]: We can clean this up a bit to both signal that we will be continuing the comment and carry the comment token:

```rust
let continue_comment_token = if config.continue_comments {
    doc.language_config()
        .and_then(|config| config.comment_tokens.as_ref())
        .and_then(|tokens| comment::get_comment_token(text, tokens, cursor_line))
} else {
    None
};
```

then in the `match` below we check if `continue_comment_token.is_some()` instead

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:3438]: style nit: we try to avoid using blocks for bindings like this. It's fine for `line` to escape into the outer scope

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:3488]: Rather than jumping to the end of the line we should update the selection ranges based on how much we're inserting in the block above with the "calculate new selection ranges" comment. I think the final formula there should be `pos + (i * (1 + indent_len + comment_len)) + indent_len + comment_len` where `comment_len` is

```rust
let comment_len = continue_comment_token.map(|token| token.len() + 1).unwrap_or_default();
```

(+ 1 for the extra space added).

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:3917]: Similar here

```rust
let continue_comment_token = if config.continue_comments {
    doc.language_config()
        .and_then(|config| config.comment_tokens.as_ref())
        .and_then(|tokens| comment::get_comment_token(text, tokens, current_line))
} else {
    None
};
```

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:3960]: We can add an extra block to special case comment continuing here:


```rust
let local_offs = if let Some(token) = continue_comment_token {
    // insert the line ending, then indent then token then ' '
} else if on_auto_pair() {
    // unchanged
} else {
    // unchanged
}
```

and switch `on_auto_pair` to be a closure that takes no arguments - that way we avoid computing it if we're continuing a comment

**Comment by the-mikedavis** [comment]
I'm not sure how much extra work it would be (maybe belongs in a follow-up PR) but it would be nice to have a way to not continue a comment. You might want to add an uncommented line after a commented line and with this change you need to go to the next line and use `O` rather than using `o` on the commented line. I think it would be handy if the first Enter (or `o`) continued the comment and then a second Enter removed the comment token and trailing `' '`.

**Comment by pascalkuthe** [comment]
> I'm not sure how much extra work it would be (maybe belongs in a follow-up PR) but it would be nice to have a way to not continue a comment. You might want to add an uncommented line after a commented line and with this change you need to go to the next line and use `O` rather than using `o` on the commented line. I think it would be handy if the first Enter (or `o`) continued the comment and then a second Enter removed the comment token and trailing `' '`.

Maybe a dedicated command/jeybinding (kaklune tends to use alt based bindings like a-o for stuff like this) you work.

This requires either adding state or doing this for every commented line which would be surprising. Also I often have empty lines in doc comments (like writing code inside markdown doc comments) so that would get annoying there

**Comment by TornaxO7** [comment]
> I'm not sure how much extra work it would be (maybe belongs in a follow-up PR) but it would be nice to have a way to not continue a comment. You might want to add an uncommented line after a commented line and with this change you need to go to the next line and use `O` rather than using `o` on the commented line. I think it would be handy if the first Enter (or `o`) continued the comment and then a second Enter removed the comment token and trailing `' '`.

I agree with pascakuthe that it might become annoying it you're trying to create an empty line inside a comment if this `Enter` behaviour is added as a mapping.

In my case, if I want to "end" a comment section, I'm just pressting `Ctrl` + `w` after creating a new comment line. Maybe that's already enough (at least it's for me).

File: helix-view/src/editor.rs
Comment (TornaxO7) [reply helix-view/src/editor.rs:350]: > I wonder if this needs to be configurable at all? I'm not sure it's configurable in Kakoune for example

hm... I kinda agree here. I thought that there could be some people who don't like to have this feature but I can't get up an idea why anyone would dislike this. I can remove this config option then. It should be simple to add it back.

File: helix-core/src/comment.rs
Comment (TornaxO7) [reply helix-core/src/comment.rs:30]: Good point! This also cleans up the code :+1:

**Comment by the-mikedavis** [comment]
Yeah good point, I hadn't thought to use C-w to remove the comment token. Kakoune's behavior for this is confusing and maybe buggy - enter sometimes strips the comment token but it depends on the language and context. Using C-w is more explicit and predictable and we should avoid ruining the experience of continuing a comment over multiple lines, for example adding an extra line to break up paragraphs in markdown in a Rust doc comment.

For that use-case specifically I believe we also want to be trimming whitespace under certain conditions in `insert_newline`, similar to https://github.com/helix-editor/helix/pull/4854 but not just when the line is all whitespace. That way you can hit `Enter` `Enter` to continue a line comment twice and trim the trailing whitespace after the first continuation. I have a change locally to do this but I'll submit it after this is merged since it will conflict (only because of formatting changes)

File: helix-core/src/comment.rs
Comment (the-mikedavis) [inline helix-core/src/comment.rs:27]:
```suggestion
        if line.byte_slice(start..).starts_with(token) {
```

one more thing I just noticed: slicing with this `end` might panic if the number of bytes of the comment token doesn't line up with the char boundaries in the text. For example if you have a line that starts with `··` in a Rust file, when you check `///` or `//!` you get a panic because the comment tokens are 3 bytes but ecah `·` is two bytes, so end lies between char boundaries

Comment (the-mikedavis) [reply helix-core/src/comment.rs:27]: (`starts_with` uses `get_byte_slice` which avoids the panic)

Comment (the-mikedavis) [reply helix-core/src/comment.rs:27]: Actually with this change we might as well remove the call to `char_to_byte` when binding `start` and use `slice(start..)` instead

Comment (TornaxO7) [reply helix-core/src/comment.rs:27]: Hm... I'm seeing a little problem here: Assuming the line comments of Rust are saved as follows:
```toml
["///", "//"]
```
due to `start._with`, `//` is going to be accepted as well, even if we are inside a `///` doc comment, so `//` would be added first.
Are you fine with also checking the length of the token?

Comment (the-mikedavis) [reply helix-core/src/comment.rs:27]: Ah good point. Instead let's use `get_byte_slice` and `is_some_and(|s| s == *token)` on the start..end byte slice. That will avoid the panic too but be a proper equality check

Comment (TornaxO7) [reply helix-core/src/comment.rs]: I implemented another solution: Is that ok for you as well?

Comment (TornaxO7) [reply helix-core/src/comment.rs]: I think that this is worth for a test (checking the length of the token).

File: helix-core/src/comment.rs
Comment (the-mikedavis) [inline helix-core/src/comment.rs:30]: This is mixing up the bytes/chars indices since `token_str.len()` is the number of bytes but we're treating it as the number of chars. It won't panic since we're capping it to the length in chars but if there were non-ASCII comment tokens there would be a correctness issue I believe.

We can use `slice(start..)` here anyways though and avoid computing `end` at all since we're using `starts_with`

Comment (the-mikedavis) [reply helix-core/src/comment.rs:30]: I believe this block can be simplified a bit using some iterator functions:

```rust
let line = text.line(line_num);
let start = line.first_non_whitespace_char()?;

tokens
    .iter()
    .map(AsRef::as_ref)
    .filter(|token| line.slice(start..).starts_with(token))
    .max_by_key(|token| token.len())
```

**Review by the-mikedavis — DISMISSED** [review]

File: helix-core/src/comment.rs
Comment (pascalkuthe) [inline helix-core/src/comment.rs:320]: This diff is pretty hard to follow. I think these can be reset together with the function I mentioned above?

File: helix-term/src/commands.rs
Comment (pascalkuthe) [inline helix-term/src/commands.rs:3954]: I don't think this is a limitation we want to have in thr lognterm but usually autopairs within comments are a bit niche and likely hard to fix until some other changss land so this is probably fine for now.

File: helix-core/src/comment.rs
Comment (TornaxO7) [reply helix-core/src/comment.rs:45]: I was used in this PR until @ the-mikedavis reviewed the code and suggested to create a new function (`get_comment_token`). When I used this function, I found a bug in this function where the text will be seen as commented if the text is empty which shouldn't be the case in my opinion.

File: helix-core/src/comment.rs
Comment (TornaxO7) [reply helix-core/src/comment.rs:320]: > This diff is pretty hard to follow.

What I'm doing here is just splitting up this one test function, which tests multiple things, into smaller tests which only test one single thing.
I see the following advantages here:
- more isolated tests
- more test functions can run in parallel => faster testing

Do you still want to merge them back into one single test function?

File: helix-core/src/comment.rs
Comment (pascalkuthe) [reply helix-core/src/comment.rs:45]: Does this lead to any observable issues with commenting? If not I would prefer to reset this change.

File: helix-core/src/comment.rs
Comment (pascalkuthe) [reply helix-core/src/comment.rs:320]: Ah I thought you made the change due to the change you made to the function above.

Taking another look it seems like a good change. I don't think the parallelism matters for these tiny tests but having multiple (named) independence tests (that can pass/fail independently) is easier to work with

File: helix-core/src/comment.rs
Comment (TornaxO7) [reply helix-core/src/comment.rs:45]: > Does this lead to any observable issues with commenting? If not I would prefer to reset this change.

For now: No, there aren't any observable issues which I encountered, only the time when I used the function. I can revert it back.

Comment (TornaxO7) [reply helix-core/src/comment.rs]: done

**Review by pascalkuthe — APPROVED** [review]

File: helix-core/src/comment.rs
Comment (TornaxO7) [reply helix-core/src/comment.rs]: Oh lol, I even added a test case for exactly this case, that's why the CI is failing. Would you still like to remove this "fix"? @pascalkuthe

Comment (TornaxO7) [reply helix-core/src/comment.rs]: Well, I just removed the test since it can be easily added back, if it's needed.

**Review by pascalkuthe — APPROVED** [review]

**Review by the-mikedavis — APPROVED** [review]

**Comment by TornaxO7** [comment]
peepoAwesome

**Comment by eanyanwu** [comment]
This just made my night. Thank you @TornaxO7 & rest of the maintainers 🙌🏾  🙌🏾  for the steady progress.
It sure pays off to read through the latest commits every few days 😆

**Comment by TornaxO7** [comment]
You're welcome
peepoBucket

**Pull request merged**
